File: app/src/main/java/com/example/app/calculator/RandomCircles.java
package com.example.app.calculator;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.view.MotionEvent;
import android.view.View;

import com.drawable.*;

import java.util.ArrayList;
import java.util.List;

import static com.util.Color.*;
import static com.util.Utility.getRandomInt;

public class RandomCircles extends View implements Runnable {
    private List<Circle> circles = new ArrayList<>();
    private Paint paint = new Paint();
    private Text[] moreCirclesText = {
            new Text("Tap the screen to add", 0, 50, 35, BLUE),
            new Text("more circles!", 0, 100, 35, BLUE)
    };
    private int width;
    private int height;
    private int speedRange = 5;

    public static final int STARTING_CIRCLES = 10;

    public RandomCircles(Context context) {
        super(context);
        for (int i = 0; i < STARTING_CIRCLES; i++) {
            circles.add(randomCircle(0, 0, speedRange));
        }
    }

    @Override
    protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight) {
        this.width = width;
        this.height = height;
        super.onSizeChanged(width, height, oldWidth, oldHeight);
        for (Circle circle : circles) {
            circle.setX(getRandomInt((int) circle.getRadius(), this.width - 60));
            circle.setY(getRandomInt((int) circle.getRadius(), this.height - 60));
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent e) {
        float x = e.getX();
        float y = e.getY();

        switch (e.getAction()) {
            case MotionEvent.ACTION_DOWN:
                circles.add(randomCircle(x, y, speedRange));
                break;
        }
        performClick();
        return super.onTouchEvent(e);
    }

    @Override
    public boolean performClick() {
        super.performClick();
        return true;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        for (Circle circle : circles) {
            circle.draw(canvas, paint);
            if (circle.getX() + circle.getRadius() > width || circle.getX() - circle.getRadius() < 0) {
                circle.setXVelocity(-circle.getXVelocity());
            }
            if (circle.getY() + circle.getRadius() > height || circle.getY() - circle.getRadius() < 0) {
                circle.setYVelocity(-circle.getYVelocity());
            }
        }
        if (moreCirclesText != null) {
            for (Text text : moreCirclesText) {
                text.draw(canvas, paint);
            }
            if (circles.size() > STARTING_CIRCLES + 10) moreCirclesText = null;
        }
    }

    private Circle randomCircle(double x, double y, int range) {
        int randXVel;
        int randYVel;
        do {
            randXVel = getRandomInt(-range, range);
        } while (randXVel > -range / 2 && randXVel < range / 2);

        do {
            randYVel = getRandomInt(-range, range);
        } while (randYVel > -range / 2 && randYVel < range / 2);

        return new Circle(x, y, getRandomInt(30, 50), getRandomColor(), randXVel, randYVel);
    }

    @SuppressWarnings({"InfiniteLoopStatement", "EmptyCatchBlock"})
    @Override
    public void run() {
        for (;;) {
            postInvalidate();
            try {
                Thread.sleep(1000 / 60); //60fps
            } catch (InterruptedException e) {}
        }
    }
}

File: app/src/main/java/com/example/app/calculator/RandomCircles.kt
package com.example.app.calculator

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.view.MotionEvent
import android.view.View

import com.drawable.*
import com.util.Color

import java.util.ArrayList

import com.util.Utility.getRandomInt

class RandomCircles(context: Context) : View(context), Runnable {
    private val circles = ArrayList<Circle>()
    private val paint = Paint()
    private var moreCirclesText: Array<Text> = arrayOf(
            Text("Tap the screen to add", 0.0, 50.0, 35.0, Color.BLUE),
            Text("more circles!", 0.0, 100.0, 35.0, Color.BLUE)
    )
    private val speedRange = 5

    init {
        for (i in 0 until STARTING_CIRCLES) {
            circles.add(Circle(0.0, 0.0, getRandomInt(30, 50).toDouble(), Color.values()[i]))
        }
    }

    override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
        super.onSizeChanged(width, height, oldWidth, oldHeight)
        for (circle in circles) {
            circle.x = getRandomInt(circle.radius.toInt(), width - 60).toDouble()
            circle.y = getRandomInt(circle.radius.toInt(), height - 60).toDouble()
            do {
                circle.xVelocity = getRandomInt(-speedRange, speedRange).toDouble()
            } while (circle.xVelocity > -speedRange / 2 && circle.xVelocity < speedRange / 2)

            do {
                circle.yVelocity = getRandomInt(-speedRange, speedRange).toDouble()
            } while (circle.yVelocity > -speedRange / 2 && circle.yVelocity < speedRange / 2)
        }
    }

    override fun onTouchEvent(e: MotionEvent): Boolean {
        val x = e.x
        val y = e.y

        when (e.action) {
            MotionEvent.ACTION_DOWN -> {
                circles.add(randomCircle(x.toDouble(), y.toDouble(), speedRange))
            }
        }
        performClick()
        return super.onTouchEvent(e)
    }

    override fun performClick(): Boolean {
        super.performClick()
        return true
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        for (circle in circles) {
            circle.draw(canvas, paint)
            if (circle.x + circle.radius > width || circle.x - circle.radius < 0) {
                circle.xVelocity = -circle.xVelocity
            }
            if (circle.y + circle.radius > height || circle.y - circle.radius < 0) {
                circle.yVelocity = -circle.yVelocity
            }
        }
        if (circles.size <= STARTING_CIRCLES + 10) {
            for (text in moreCirclesText) {
                text.draw(canvas, paint)
            }
        }
    }

    private fun randomCircle(x: Double, y: Double, range: Int): Circle {
        var randXVel: Int
        var randYVel: Int
        do {
            randXVel = getRandomInt(-range, range)
        } while (randXVel > -range / 2 && randXVel < range / 2)

        do {
            randYVel = getRandomInt(-range, range)
        } while (randYVel > -range / 2 && randYVel < range / 2)

        return Circle(x, y, getRandomInt(30, 50).toDouble(), Color.getRandomColor(), randXVel.toDouble(), randYVel.toDouble())
    }

    override fun run() {
        while (true) {
            postInvalidate()
            try {
                Thread.sleep((1000 / 60).toLong())
            } catch (e: InterruptedException) {}
        }
    }

    companion object {
        val STARTING_CIRCLES = Color.valueOf("CYAN").ordinal + 1
    }
}